### ISSUE TITLE
Account data structure refactoring for HDKD support

### ISSUE THREAD

**Issue by Tbaut** [paritytech/parity-signer#419]
Edited with Gav's suggestions.

`AccountStore` should be refactored to have the following elements:

Should look like the following:
```js
identities = [{
    // first identity
    encryptedMnemonic : 'report eager... gossip fury', //this should be encrypted with a pin
    networks : {
        "//kusama-cc1": {
            meta : {
                "//funding/0": {
                    nick "My Kusama account",
                    address: "GnuqPFhJsvv2NDHqoZeGPPFS8yxTmMzZYKTuc1vCFYYT9fu"
                },
                "//funding/1": {
                    address: "vv2NDhJsvv2NDHqoZeGPPFS8yxTmMzZYKTuc1vCFYYTYT9"
                }
            }
            addresses : {
                "GnuqPFhJsvv2NDHqoZeGPPFS8yxTmMzZYKTuc1vCFYYT9fu": "//funding/0",
                "vv2NDhJsvv2NDHqoZeGPPFS8yxTmMzZYKTuc1vCFYYTYT9": "//funding/1"
            }
        },
        "//polkadot": {
            meta : {
                "//stash/0": {
                    nick "My Polkadot stash",
                    address: "hJsvPFhJsvv2NDHqoZeGPPFS8yxTmMzZYKTuc1vCFYYT9fu"
                },
                "//controller/0": {
                    address: "eGPPDhJsvv2NDHqoZeGPPFS8yxTmMzZYKTuc1vCFYYTYT9"
                }
            }
            addresses: 
                "hJsvPFhJsvv2NDHqoZeGPPFS8yxTmMzZYKTuc1vCFYYT9fu": "//stash/0",
                "eGPPDhJsvv2NDHqoZeGPPFS8yxTmMzZYKTuc1vCFYYTYT9": "//controller/0"
            }
        }
    }
},{
    // second identity
    ...
}]
```

How to go from what the QR code gives us to an identity and path:
```js
let genesis, address
// find the network and its prefix based on the genesis hash
let network = networks.find(n => n.genesis == genesis) || networks["unknown"]
// get the address based on the scanned public key and the network prefix
// find the identity that contains this address (O(n) but `identities` will remain tiny)
let identity = identities.find(identity => identity[network.name].addresses[address])
// from there we have the identity and the full path
```

- The network mapping should remain the same as it currently is.
- We need to make sure there's a migration from the old data structure to the new one. This would result in creating one identity per old account.

**Comment by yjkimjunior** [comment]
rather than `[0]: {}, etc.`, why not an array?

**Comment by Tbaut** [comment]
I think the idea from @maciejhirsz was that if you delete an account, the indexing will not change,
 `[0]: {},[1]: {},[2]: {}` and you delete 1, you end up with `[0]: {},[2]: {}` and your indexing doesn't change. get(2) will still give you the 3rd account, although it's in second position now.

If you had an array, and say a mapping between an account id and its index, deleting things would shift things around and you'd have to recreated your indexes.

**Comment by hanwencheng** [comment]
IMO, javascript array is just special form of object with pre-defined prototypes. When we delete the item in array, it use an `empty` slot as the placeholder in the object. 

Effect is that some array method is not changed, like `arr.length` is still 3, though the object related method still have the correct result like `Object.keys(arr)  === ["0", "2"]`. 

So I would suggest let us avoid the above difference and use Object instead.

**Comment by hanwencheng** [comment]
A question: should we prevent the same network key like `kusama` with different genesis Hash?

Imagine we have the networkSpecs array like:
```javascript
const NetworkSpecs = [{
	networkKey: 'kusama',
	prefix: 2,
	genesisHash: 'abc',
	networkName: 'Kusama CC2',
}, {
	networkKey: 'kusama',
	prefix: 2,
	genesisHash: 'efg',
	networkName: 'Kusama CC3',
}]
```
So if user have both `//funding//default` accounts under these two network, in networkListScreen after they clicked they will have the account with same path which are `//kusama//funding//default`. And the only different hint is the identicon.

**Comment by Tbaut** [comment]
> I understand that, but I was thinking we could just use delete?

Interesting, I didn't know about `delete` on arrays.

> So I would suggest let us avoid the above difference and use Object instead.

Do you mean let's use object mapping, like described in the first post? I have no strong opinion. @maciejhirsz ?

>Imagine we have the networkSpecs array like
The names should be different and the derivation path should be //kusamacc1, //kusamacc2 to avoid confusion.

**Comment by sjeohp** [comment]
Does the decrypted seed need to exist on the root account?  I think it would be better to decrypt the seed for specific tasks, delete it immediately, and do away with mutable account methods where possible (`deleteSensitiveData`, `unlockAccount`).  We could then treat the account as a persistent structure.  So something like

`unlockAccount(account);`
`doSomething(account.seed);`
`...`
`deleteSensitiveData(account);`

would just become:

`doSensitive((seed) => {
...
});` 

where `doSensitive` looks like
`function doSensitive(func) {  const seed = decrypt(encryptedSeed);  func(seed);  seed = null;  }`

Which doesn't completely guarantee the seed will be removed from memory but at least makes it easier to keep track of, and means we don't need to worry about the order different mutable methods are called in.

**Comment by maciejhirsz** [comment]
Re array vs map: both are viable options, but you have to keep in mind that with an array you have to keep a tombstone entry in the array (be it `null`, `undefined`, or something) to avoid shifting things, you have to serialize those tombstones when you write to disc, and filter them out when listing all elements.

It's easier to add stuff to array since you can just get `length` as your next index and then `push`, but on map you can store the next index as a special property like `_autoIndex: 0`. You need to increment that manually, but that can be easily abstracted if you build a class that effectively acts as a database table.

I usually prefer to use arrays where I can use arrays since they are dumber, simpler, and faster (although both are O(1)),  but an ever-growing array with empty slots that stay empty forever feels dirty to me. There are ways to fix that by reusing slots, but that requires you to either track empty slots in another array (complexity) or searching for them when adding stuff (making it O(n)), and you have to be absolutely sure that any references to previously deleted items elsewhere have been purged before an old index is assigned to a new entry, and any bug here will result in bizzare bugs that are hard to reproduce and debug.

Bottom line is, whoever implements this should choose whatever they think is best, but a good rule of thumb I'd use here is "which way will require me to write less complex code", since as we all know, code is the principal source of bugs :).

**Comment by hanwencheng** [comment]
closed by #431
